python/ray/dag/compiled_dag_node.py

```diff
@@ -284,7 +308,46 @@ def __init__(
 
         self.resolved_inputs: List[Union[Any, ChannelInterface]] = []
         self.input_channels: List[ChannelInterface] = []
-        self.input_channel_idxs: List[int] = []
+        # The inner list is indices of resolved_args that correspond
```

Comment:
The logic here is a little bit hard to follow. I wonder if we can change to something more like this?

```python
# represents an arg read from a channel.
# (index into self.input_channels, index into self.resolved_inputs, arg / kwarg / None to use the value directly)
ChannelInput = Tuple[int, int, Optional[Union[int, str]]]

self.channel_inputs: List[ChannelInput]
```

It would also be great to consolidate the logic to build the channel inputs into the ExecutableTask constructor, and move input resolution into a method. That way we can make these "private" (or as private as Python lets us :) )

Reply:
Good suggestion. I refactored the code and introduced `ExecutableTaskInput`.